lib/ruby-push-notifications/wns/wns_connection.rb

```diff
@@ -0,0 +1,129 @@
+      # (http://msdn.microsoft.com/pt-br/library/windows/apps/ff941099)
+      def self.post(notifications, access_token, options = {})
+        body = notifications.as_wns_xml
+        headers = build_headers(access_token, notifications.data[:type], body.length.to_s)
+        responses = []
+        notifications.each_device do |url|
+          http = Net::HTTP.new url.host, url.port
+          http.open_timeout = options.fetch(:open_timeout, 30)
+          http.read_timeout = options.fetch(:read_timeout, 30)
+          if url.scheme == 'https'
+            http.use_ssl = true
+            http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+          end
+          response = http.post url.request_uri, body, headers
+          responses << { device_url: url.to_s, headers: extract_headers(response), code: response.code.to_i }
+        end
+        WNSResponse.new responses
+      end
+
+      # Get access auth token for sending pushes
+      # You can get it on https://account.live.com/developers/applications/index
+      #
+      # @param type [String]. Sid
+      # @param type [String]. Secret
+      #
+      # https://docs.microsoft.com/en-us/windows/uwp/controls-and-patterns/tiles-and-notifications-windows-push-notification-services--wns--overview
+      def self.access_token(sid, secret)
+        body = {
+          grant_type: 'client_credentials',
+          client_id: sid,
+          client_secret: secret,
+          scope: 'notify.windows.com'
+        }
+
+        url = URI.parse "https://login.live.com/accesstoken.srf"
+        http = Net::HTTP.new url.host, url.port
+        http.use_ssl = true
+        http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+        response = http.post url.request_uri, URI.encode_www_form(body)
+        response = JSON.parse(response.body)
+
+        if response['error']
+          raise WNSGetTokenError, response['error_description']
+        else
+          response['access_token']
+        end
+      end
+
+      # Build Header based on type and delay
+      #
+      # @param type [String]. Access token
+      # @param type [Symbol]. The type of notification
+      # @param type [String]. Content length of body
+      # @return [Hash]. Correct delay based on notification type
+      # https://msdn.microsoft.com/en-us/library/windows/apps/hh465435.aspx#send_notification_request
+      def self.build_headers(access_token, type, body_length)
+        {
+          CONTENT_TYPE_HEADER => CONTENT_TYPE[type],
+          X_WNS_TYPE_HEADER => WP_TARGETS[type],
+          CONTENT_LENGTH_HEADER => body_length,
+          AUTHORIZATION_HEADER => "Bearer #{access_token}",
+          REQUEST_FOR_STATUS_HEADER => 'true'
+        }
+      end
+
+      # Extract headers from response
+      # @param response [Net::HTTPResponse]. HTTP response for request
+      #
+      # @return [Hash]. Hash with headers with case-insensitive keys and string values
+      def self.extract_headers(response)
```

Comment:
Rather than extracting, this method is actually capitalising the headers. `capitalize_headers` would be a better name I guess.

Reply:
you are right, it will be better naming